internal/gatewayapi/translator.go

```diff
@@ -777,6 +899,71 @@ func isValidCrossNamespaceRef(from crossNamespaceFrom, to crossNamespaceTo, refe
 	return false
 }
 
+// Checks if a hostname is valid according to RFC 1123 and gateway API's requirement that it not be an IP address
+func isValidHostname(hostname string) (valid bool, reason string) {
+	if len(hostname) == 0 || hostname == "" {
+		return false, "Hostname cannot be empty"
+	}
+
+	if len(hostname) > 255 {
+		return false, fmt.Sprintf("Hostname '%s' has a length of %d. cannot exceed 255", hostname, len(hostname))
+	}
+
+	if hostname[0] == '-' || hostname[0] == '.' {
+		return false, "Hostname cannot start with '.' or '-'"
+	}
+	if hostname[len(hostname)-1] == '-' || hostname[len(hostname)-1] == '.' {
+		return false, "Hostname cannot end with '.' or '-'"
+	}
+
+	// The only allowed characters in hostnames are a-z, A-Z, 0-9, ., and -
+	var validChar = regexp.MustCompile(`[a-zA-Z0-9.-]`).MatchString
```

Comment:
My gut reaction is that regexes are big and slow, and `strings.Contains("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.-", char)` would be comparatively faster.

Reply:
Switched to `IsDNS1123Subdomain` which covers most of these checks.